### DISCUSSION TITLE
Type not narrowed correctly in else branches?

### DISCUSSION THREAD

**Discussion by crides** [microsoft/pyright#6095] [General]
For the following code:
```python
from typing import cast, assert_type

a = cast(list[int] | int, 3)
if type(a) == list:
    assert_type(a, list[int])
else:
    assert_type(a, int)
```

Somehow pyright can infer that `a` is indeed `list[int]` in the true branch, but in the else branch `a` is inferred as `list[int] | int` still. Is this expected behavior?

**Comment by erictraut** [comment]
Yes, this is expected behavior. It's unsafe to narrow the type in the negative (else) case because you're using the `type(a) == x` type guard instead of `isinstance(a, x)`. This example shows why this is unsafe:

```python
def func(a: list[int] | int):
    if type(a) == list:
        print("Got a list!")
    else:
        print("Got an int?")

class MyList(list[int]):
    ...

func(MyList()) # Prints "Got an int?"
```

Not surprisingly, mypy produces the same type narrowing results as pyright in this case.

**Reply by crides** [reply]
Ah, because `type(...) == ...` checks for the "current" type, not the "tree" of subtypes.

By the way, is there a way to check if `a` is actually `list[int]` (or any subclass of it)?

**Reply by erictraut** [reply]
> By the way, is there a way to check if a is actually list[int] (or any subclass of it)?

```python
def is_list_of_ints(val: object) -> TypeGuard[list[int]]:
    return isinstance(val, list) and all(isinstance(x, int) for x in val)
```

**Reply by crides** [reply]
I see. Much thanks for the speedy answers!

**Comment by paulzzy** [comment]
Apologies for reviving a necrothread, but here's a method that works for type variables (like `type[Foo]`). For some reason `isinstance` doesn't work (I get errors talking about `<subclass of type Foo and Bar>`).

```py
def is_type_foo(type_) -> TypeGuard[type[Foo]]:
    return type_ is Foo or issubclass(type_, Foo)
```
